File: csharp/890.find-and-replace-pattern.cs
/*
 * @lc app=leetcode id=890 lang=csharp
 *
 * [890] Find and Replace Pattern
 */

using System.Collections.Generic;
using System.Linq;

// @lc code=start
public class Solution {

    private static bool IsMatch(string word, string pattern) {
        var w2p = new Dictionary<char, char>();
        var ps = new HashSet<char>();
        int n = pattern.Length;
        if (n != word.Length) {
            return false;
        }
        for (int i = 0; i < n; ++i) {
            char w = word[i];
            if (w2p.TryGetValue(w, out char p)) {
                if (p != pattern[i]) {
                    return false;
                }
                continue;
            }
            if (!ps.Add(pattern[i])) {
                return false;
            }
            w2p[w] = pattern[i];
        }
        return true;
    }

    public IList<string> FindAndReplacePattern(string[] words, string pattern) {
        return words
            .Where(word => IsMatch(word, pattern))
            .ToList();
    }
}
// @lc code=end

File: java/890.find-and-replace-pattern.java
/*
 * @lc app=leetcode id=890 lang=java
 *
 * [890] Find and Replace Pattern
 */

// @lc code=start
class Solution {
    public List<String> findAndReplacePattern(String[] words, String pattern) {
        List<String> ret = new ArrayList<String>();
        int n = pattern.length();
        for (String word : words) {
            int[] wtp = new int[26], ptw = new int[26];
            boolean flag = false;
            for (int i = 0; i < n && !flag; ++i) {
                int w = (int)word.charAt(i) - (int)'a', p = pattern.charAt(i) - (int)'a';
                if (wtp[w] == 0) {
                    if (ptw[p] != 0) {
                        flag = true;
                    } else {
                        wtp[w] = p + 1;
                        ptw[p] = w + 1;
                    }
                } else {
                    if (wtp[w] != p + 1) {
                        flag = true;
                    }
                }
            }
            if (!flag) {
                ret.add(word);
            }
        }
        return ret;
    }
}
// @lc code=end

